Fix Focus Punch/Shell Trap behavior with Encore

scheibo opened the pull request:
> - [x] Encore should work successfully on Shell Trap, in a similar way to the Focus Punch behavior. EDIT: see [here ](https://www.smogon.com/forums/threads/bug-reports-v3-0-read-op-before-posting.3634749/post-7796738)for clarification.

_Originally posted by @DaWoblefet in https://github.com/Zarel/Pokemon-Showdown/issues/2367#issuecomment-390426966_

@Marty-D PTAL :)

scheibo commented:
Actually, I think I missed this?

> If a user's previous move was Focus Punch / Shell Trap and it is Encored into Focus Punch / Shell Trap while attempting to execute the move, the regular "you can't be hit" effect for Focus Punch and the "you must be hit" effect of Shell Trap will be enforced.

scheibo commented:
> Actually, I think I missed this?

Hmm, actually after updating the tests such that I think they actually test what @DaWoblefet was explaining my fix appears to cover it?

Marty-D commented:
Most of the time DaWoblefet is referring to Focus Punch/Shell Trap being Encored while the target was trying to use a different move. The "always works/never works" condition is only relevant to the turn that Pokemon's decision was changed by Encore. Other turns of Encore have Focus Punch/Shell Trap work normally. You have to actually click the move on your screen and have them "charge" to have their full effects happen.

scheibo commented:
Thanks for the explanation - I've updated the tests and reverted my 'fix' - it seems like Shell Trap always failing is the only thing that is failing which makes sense because "in a similar way to the Focus Punch behavior." seems to imply Focus Punch was already working.

Marty-D commented:
> - [x] Encore should work successfully on Shell Trap, in a similar way to the Focus Punch behavior. EDIT: see [here ](https://www.smogon.com/forums/threads/bug-reports-v3-0-read-op-before-posting.3634749/post-7796738)for clarification.

This bit is referring to Encore working on Shell Trap at all. Currently on PS, Focus Punch and Shell Trap only count as the last move used when they're successful, which happens more often for Focus Punch which is why it seems like Encore works on it but not Shell Trap. In-game they always count as the user's last move, and are subject to being Encored after that.

Also probably related: Shell Trap should use PP whether it's successful or not, while Focus Punch only uses PP when it's successful (which only started happening in Gen 5). I believe I tried fixing this part several months ago but something got in the way, and I feel like it was because Shell Trap is a spread move.

scheibo commented:
> Currently on PS, Focus Punch and Shell Trap only count as the last move used when they're successful, which happens more often for Focus Punch which is why it seems like Encore works on it but not Shell Trap

Ah! That's what was missing from my test - I've fixed it so now they *both* fail. Thanks!

> Also probably related: Shell Trap should use PP whether it's successful or not, while Focus Punch only uses PP when it's successful (which only started happening in Gen 5). I believe I tried fixing this part several months ago but something got in the way, and I feel like it was because Shell Trap is a spread move.

I've added tests for these as well. The tests seem to imply the Focus Punch behavior as implemented today is correct?

I'll take a stab at fixing these if you can confirm the tests look like they're correctly outlining the desired behavior. If I can't figure it out I'll at least submit the tests (with `it.skip`) to make the next person's job easier.

Marty-D commented on data/moves.js:
```diff
@@ -14757,10 +14757,11 @@ let BattleMovedex = {
 			pokemon.addVolatile('shelltrap');
 		},
-		beforeMoveCallback(pokemon) {
+		onTry(pokemon) {
```
Oh yeah! This was exactly what I tried when I attempted to fix it and it failed spectacularly because `Try` runs against each target, so when Shell Trap succeeds or fails it does so multiple times.

MacChaeger replied:
Just sayin', #5216 fixes that.

scheibo replied:
@MacChaeger - fixes it so that 0962c84 would actually work and not have the problems @Marty-D describes?

MacChaeger replied:
Well, events like `Try` only run against one target in my version.

scheibo replied:
Gotcha. I tried to repro the '`Try` runs against each target' bug Marty described in 034fbac / c06192a, but the test still seems to pass? It seems like Shell Trap is only succeeding once, at least according to the damage its doing and the amount of PP its losing?

Marty-D replied:
Hmm, maybe I misremembered the success part since it doesn't happen twice currently, but `cant` is indeed sent twice when it fails.

scheibo replied:
I've modified the test to check the log for the double `cant`. Just to verify: https://gist.github.com/scheibo/c9a2a78708420b53f93678a8370fd2a6 is the log after the second turn, where `cant` shows up twice - does it look invalid?

I'm going to leave the move with `beforeMoveCallback` with a TODO for the time being (and the test `skip`ed) - once @MacChaeger lands his spread move fix we can switch to using `onTry` and reenable the test.

Marty-D commented:
>In-game they always count as the user's last move, and are subject to being Encored after that.
Sorry, I don't know why I said this; Focus Punch can't be Encored if it failed, but Shell Trap can. And they can both be Encored if they succeeded.

Marty-D commented on test/simulator/moves/encore.js:
```diff
@@ -0,0 +1,124 @@
+		// If a user's previous move was Focus Punch and it is Encored into Focus Punch while attempting to
+		// execute the move, the regular "you can't be hit" effect for Focus Punch will be enforced.
+		battle.makeChoices('move focuspunch 1, move teleport', 'move encore 1, move extremespeed 1');
+		assert.strictEqual(hp, battle.p2.active[0].hp);
```
You'll have to make the previous turn a successful Focus Punch to get this to work, sorry about that!

Marty-D commented on test/simulator/moves/encore.js:
```diff
@@ -0,0 +1,124 @@
+		// if it is hit at any time before or after the Encore, it still uses Focus Punch like normal. It doesn't matter
+		// in the case of Focus Punch if the user was hit before or after the Encore; Focus Punch will still always work.
+		battle.makeChoices('move splash, move teleport', 'move encore 1, move extremespeed 1');
+		assert.notStrictEqual(hp, battle.p2.active[0].hp);
```
You'll have to make the previous turn a successful Focus Punch to get this to work, sorry about that!

scheibo replied:
> Sorry, I don't know why I said this; Focus Punch can't be Encored if it failed, but Shell Trap can. And they can both be Encored if they succeeded.

No worries, I'm just in awe that you guys know all this! 

> You'll have to make the previous turn a successful Focus Punch to get this to work, sorry about that!

Ah, OK. So changing this back would cause the test to pass, indicating Focus Punch is in fact implemented correctly in PS today. Is that right?

Marty-D replied:
Yeah, Focus Punch behaviour is currently correct as far as I know.

Marty-D commented on data/moves.js:
```diff
@@ -14757,10 +14757,11 @@ let BattleMovedex = {
 		},
-		beforeMoveCallback(pokemon) {
+		onTry(pokemon) {
 			if (pokemon.volatiles['shelltrap'] && !pokemon.volatiles['shelltrap'].gotHit) {
```
```suggestion
			if (!pokemon.volatiles['shelltrap'] || !pokemon.volatiles['shelltrap'].gotHit) {
```
This should make Shell Trap fail correctly when forced to be used from a decision-changing Encore.

scheibo replied:
This didn't seem to fix things? But maybe my test is wrong.

scheibo replied:
My test was wrong. Thanks for the fix!

Marty-D commented:
I believe the final failing test(s) after the suggested change are related to Encore maybe happening before Splash in the second last Shell Trap test. Or the attempted use of Focus Punch in the last test :P

Marty-D commented on data/moves.js:
```diff
@@ -14757,10 +14757,11 @@ let BattleMovedex = {
-		beforeMoveCallback(pokemon) {
+		onTry(pokemon) {
 			if (pokemon.volatiles['shelltrap'] && !pokemon.volatiles['shelltrap'].gotHit) {
+				this.attrLastMove('[still]');
```
Is this necessary? I think `Try` runs before the animation plays.

scheibo replied:
I copied it from Focus Punch. Shoudl that be updated as well?

Marty-D replied:
Oh that's right, it does play because 'POKEMON used MOVE!' happens, which lets the client know PP was deducted. This is fine then, my bad.

scheibo commented:
> I believe the final failing test(s) after the suggested change are related to Encore maybe happening after Splash in the second last Shell Trap test. 

Can you elaborate here? I currently have 2 failing tests (both Encore + Shell Trap ones) and 1 ignored (the Shell Trap PP one, because it depends on @MacChaeger). I would expect the first Shell Trap one ("it should not affect Shell Trap if the user's decision is not changed") to work, but don't see how Splash effects things? 

> Or the attempted use of Focus Punch in the last test

Doh. Fixed.

I expect this last test to fail (this is the new behavior I want to dd), and the change you proposed doesn't seem to fix it (regardless of whether we're using `onTry` or `beforeMoveCallback`)

Marty-D commented:
If the Smeargle with Encore moves after the Smeargle with Splash, Shell Trap won't be used that turn.

scheibo commented:
> If the Smeargle with Encore moves after the Smeargle with Splash, Shell Trap won't be used that turn, or the next turn

Both Smeargles have Splash, so I assume you mean the Smeargle with Encore vs. the Smeargle with Shell Trap?

> The Zigzagoon may have fainted at some point from Shell Trap damage and can't use Extreme Speed?

This shouldn't be the case, I had already calced this:

`0 SpA Smeargle Shell Trap vs. 0 HP / 0 SpD Smeargle: 66-78 (26.2 - 31%) -- guaranteed 4HKO`
`0 SpA Smeargle Shell Trap vs. 0 HP / 0 SpD Zigzagoon: 70-83 (32.2 - 38.2%) -- 97.1% chance to 3HKO`

Though I guess I didn't account for crits, which could make the test flaky (but that isn't the case here). However, you're right about the fainting - espeed is power and it KOs the Smeargle using Shell Trap! I guess I forgot that I made Zigzagoon Level 100 in this test to survive the shell traps. 

---

With the tests fixed so that only the "Encore: should make Shell Trap always fail if the user's decision is changed" is the only one failing, once I add in your change it works. I think this change is now ready to go!

Marty-D commented on test/simulator/moves/encore.js:
```diff
@@ -0,0 +1,123 @@
+		// attack, and will always say "<Pokemon>'s shell trap didn't work. It doesn't matter in the case of Shell Trap if
+		// the user was hit before or after the Encore; Shell Trap will still always fail.
+		battle.makeChoices('move splash, move teleport', 'move encore 1, move quickattack 1');
+		assert.strictEqual(battle.p2.active[0].hp, hp);
```
Without the change to `onTry` for Shell Trap, the only reason this works is because Encore fails against the failed Shell Trap, and Splash predictably deals no damage to the other Smeargle so the test passes. Do we want to just leave it like this until spread move order is fixed?
